### PR TITLE
[WIP] POC for Index deprecation enforcements

### PR DESCRIPTION
**Pull request by galipremsagar**
## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/index.py [hunk 1]
```diff
@@ -69,8 +66,39 @@
...
+    """Custom metaclass for Index that overrides instance/subclass tests."""
+
+    def __call__(cls, data, *args, **kwargs):
+        if cls is Index:
+            return as_index(
+                arbitrary=data,
+                *args,
+                **kwargs,
+            )
+        return super().__call__(data, *args, **kwargs)
```
[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/core/index.py:79]: Your idea, you mentioned this while we were chatting about it. 😉

File: python/cudf/cudf/core/index.py [hunk 2]
```diff
@@ -920,7 +943,7 @@ def __abs__(self):
         return abs(self._as_int_index())
 
 
-class GenericIndex(SingleColumnFrame, BaseIndex):
+class Index(SingleColumnFrame, BaseIndex, metaclass=IndexMeta):
```
[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/core/index.py:948]: Like?

[the remark this answers is not in the document]
Comment (vyasr) [reply python/cudf/cudf/core/index.py:948]: Yeah after our meeting I see what you're going for here and this is fine.

File: python/cudf/cudf/core/index.py [hunk 3]
```diff
@@ -69,8 +66,41 @@
...
+
+    def __instancecheck__(self, instance):
+        if self in {
+            cudf.CategoricalIndex,
+            cudf.TimedeltaIndex,
+            cudf.DatetimeIndex,
+            cudf.IntervalIndex,
+        }:
+            return False
+        return isinstance(instance, BaseIndex)
```
Comment (vyasr) [inline python/cudf/cudf/core/index.py:89]: Would it be simpler to invert the condition? `if self is cudf.Index` rather than having to list all the different child classes?

File: python/cudf/cudf/core/multiindex.py [hunk 4]
```diff
@@ -1007,10 +1007,15 @@ def _concat(cls, objs):
...
-                names[i] = names[i] or name
+        objs = set(
+            map(lambda x: x.names, filter(lambda o: o.names is not None, objs))
+        )
+        if len(objs) == 1:
+            # Only set names if all `objs` have same `names`,
+            # else ignore.
+            for o in objs:
+                for i, name in enumerate(o):
+                    names[i] = names[i] or name
```
Comment (wence-) [inline python/cudf/cudf/core/multiindex.py:1018]: This logic is kind of hard to follow, how about:
```python
try:
    # Only set names if all objs have the same names
    (names,) = {o.names for o in obj} - {None}
except ValueError:
   names = [None] * source_data._num_columns
```

Note that names is initialised to be a list of `None` on line 1009, so the `names[i] or name` check only ever produces `name`.

File: python/cudf/cudf/testing/testing.py [hunk 5]
```diff
@@ -65,25 +65,15 @@ def _check_types(
...
-                right,
-                (
-                    cudf.Int8Index,
-                    cudf.Int16Index,
-                    cudf.Int32Index,
-                    cudf.Int64Index,
-                ),
+            and (
+                isinstance(right, cudf.Index)
+                and str(right.dtype) in {"int8", "int16", "int32", "int64"}
```
Comment (wence-) [inline python/cudf/cudf/testing/testing.py:70]:
```suggestion
                and right.dtype.kind == "i"
```
?

File: python/cudf/cudf/testing/testing.py [hunk 6]
```diff
@@ -65,25 +65,15 @@ def _check_types(
...
-                left,
-                (
-                    cudf.Int8Index,
-                    cudf.Int16Index,
-                    cudf.Int32Index,
-                    cudf.Int64Index,
-                ),
+            and (
+                isinstance(left, cudf.Index)
+                and str(left.dtype) in {"int8", "int16", "int32", "int64"}
```
Comment (wence-) [inline python/cudf/cudf/testing/testing.py:76]:
```suggestion
                and left.dtype.kind == "i"
```
?

File: python/cudf/cudf/core/multiindex.py (hunk 4 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/multiindex.py]: Done

File: python/cudf/cudf/core/index.py [hunk 7]
```diff
@@ -988,24 +992,42 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
...
-                if type(o) is GenericIndex and o.dtype.kind == "b":
+                # boolean Indexes, not dtype-specific subclasses.
+                if type(o) is Index and o.dtype.kind == "b":
                     out[i] = o.values
 
             return out[0] if ufunc.nout == 1 else tuple(out)
 
         return NotImplemented
 
+    @copy_docstring(StringMethods)  # type: ignore
```
[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/core/index.py:1000]: It's not specific for `copy_docstring`, any decorator added to `property` isn't supported by `mypy`:https://github.com/python/mypy/issues/1362

```
error: Decorators on top of @property are not supported  [misc]
```

To ignore this we had to add `type: ignore`

File: python/cudf/cudf/core/index.py (hunk 3 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/index.py]: Done 👍

File: python/cudf/cudf/testing/testing.py (hunk 5 above)
Comment (galipremsagar) [reply python/cudf/cudf/testing/testing.py]: Done 👍

File: python/cudf/cudf/core/index.py [hunk 8]
```diff
@@ -2257,7 +2257,7 @@ def days(self):
         Number of days for each element.
         """
         return as_index(
-            arbitrary=self._values.days, name=self.name, dtype="int32"
+            arbitrary=self._values.days, name=self.name, dtype="int64"
```
Comment (wence-) [inline python/cudf/cudf/core/index.py:2260]: Why does only this one need changed to int64?

Comment (galipremsagar) [reply python/cudf/cudf/core/index.py:2260]: We had to change just for `days` because it is not constrained like the rest of parameters in the range of values and there is an overflow happening:
```python
In [1]: import cudf

In [2]: i = cudf.Index([245345345545332], dtype='timedelta64[s]')

In [3]: i.days
Out[3]: Index([2839645203], dtype='int64')

In [4]: i
Out[4]: TimedeltaIndex(['2839645203 days 01:42:12'], dtype='timedelta64[s]')

In [5]: i.days
Out[5]: Index([-1455322093], dtype='int32')          # Overflow prior to this change.
```

File: docs/cudf/source/developer_guide/library_design.md [hunk 9]
```diff
@@ -89,12 +89,12 @@ While we've highlighted some exceptional cases of Indexes before, let's start wi
...
 
-Almost all indexes are subclasses of `GenericIndex`, a single-columned index with the class hierarchy:
+Almost all indexes are subclasses of `Index`, a single-columned index with the class hierarchy:
 ```python
-class GenericIndex(SingleColumnFrame, BaseIndex)
+class Index(SingleColumnFrame, BaseIndex)
 ```
 Integer, float, or string indexes are all composed of a single column of data.
-Most `GenericIndex` methods are inherited from `Frame`, saving us the trouble of rewriting them.
+Most `Index` methods are inherited from `Frame`, saving us the trouble of rewriting them.
```
[the remark this answers is not in the document]
Comment (galipremsagar) [reply docs/cudf/source/developer_guide/library_design.md:97]: Good catch. Updated it. Let me know if you think it needs more clarification.

File: python/cudf/cudf/core/index.py (hunk 8 above)
Comment (wence-) [reply python/cudf/cudf/core/index.py:2265]: Aha, thanks! Maybe add a comment?

Comment (galipremsagar) [reply python/cudf/cudf/core/index.py:2266]: Done 👍

**Review by wence- — APPROVED** [review]
Thanks @galipremsagar

**Comment by bdice** [comment]
@galipremsagar Do any of these fixes need backported to `branch-23.08` for pandas 1.x compatibility, or is that unnecessary?
